packages/react-instantsearch-core/src/lib/useWidget.ts

```diff
@@ -83,7 +88,11 @@ export function useWidget<TWidget extends Widget | IndexWidget, TProps>({
     };
   }, [parentIndex, widget, shouldAddWidgetEarly, search, props]);
 
-  if (shouldAddWidgetEarly) {
+  if (shouldAddWidgetEarly && promiseRef.current?.status === 'pending') {
     parentIndex.addWidgets([widget]);
   }
+
+  if (typeof window === 'undefined' && promiseRef.current) {
```

Comment:
in the future we'll need to rename promiseRef, as it's not too clear what it's goal is